lang/llvm-3.4/files/0004-Set-the-Mach-O-CPU-Subtype-to-ppc7400-when-targeting.patch

```diff
@@ -32,7 +32,7 @@ index 0d42081..7260c10 100644
        bool is64 = getPointerSize() == 8;
 +      MachO::CPUSubTypePowerPC CS = MachO::CPU_SUBTYPE_POWERPC_ALL;
 +
-+      if (TheTriple.isOSDarwin() && !TheTriple.isMacOSXVersionLT(10,5)) {
++      if (TheTriple.isOSDarwin() && (!TheTriple.isMacOSXVersionLT(10,5) && !TheTriple.isMacOSXVersionLT(10,6))) {
```

Comment:
Why not `isMacOSXVersionLT(10,6)` only?

Reply:
@catap Fair enough.

I will test the build again, if all goes as planned, will update the PR and hopefully it gets pushed. `llvm34` is a big stopped on Rosetta otherwise.

Reply:
well, even if it builds, it doesn't work :P

I think we just need to get llvm out of the variants for cctools and ld64 on PowerPC, which would be by far the better option IMHO. I believe that is the only reason it is wanted at all anyway.

Of course, someday, over the rainbow, way up high, iain will maybe actually fix llvm and clang for powerpc and then some of this llvm stuff starts to make sense.

Reply:
@kencu I feel that migration to xtools seems much early :)

Reply:
> well, even if it builds, it doesn't work :P

@kencu It works for what it is currently used for, which some LTO magic or w/e.

> I think we just need to get llvm out of the variants for cctools and ld64 on PowerPC, which would be by far the better option IMHO. I believe that is the only reason it is wanted at all anyway.

Unfortunately, `xcode` variant of `ld64` is broken, so until we move to `xtools`, there is no other feasible option. (Well, unless someone fixes `cctools` properly, but that cannot be myself.)

Reply:
xtools won't really help this llvm34 issue, but someday, after a LOT LOT LOT of testing, MacPorts might consider moving whole hog to Iain's toolchain, on all systems where it is newer than what we have. What I am running now on 10.6 Intel is already MUCH newer than what he has, so I will never change to xtools. Maybe 10.4 and 10.5 might change, and @barracuda156 's 10.6-on-PowerPC passion. There will be LOTS of discussion about that, I bet :>

the xcode variant of ld64 was never meant to be used on any systems except the very very newest ones (10.15 and newer) so you should just stay well away from that on older systems. It is of absolutely no value anyway, as the cctools and ld64 in those ancient xcodes is from the 2008 era. Why would you want that?????

I use (and build) cctools and ld64 just fine on every system from 10.4 PPC to Monterey. What exact problem are you having that you are trying to fix? I think it is that one or both of cctools and ld64 defaults to invoking LLVM to get LTO, all of which is pretty much useless on PowerPC. The fix there is to make it stop asking for LLVM in the first place. Anyway, llvm34 builds on 10.5 PPC as I recall.

Don't mess things up too much while you're playing around please,, guys. This has all worked for years, so please don't blow it all up to support Rosetta on 10.6....play nice :>

Reply:
> The fix there is to make it stop asking for LLVM in the first place.

@kencu Likely so, but who is gonna make it?

> Anyway, llvm34 builds on 10.5 PPC as I recall.

It builds fine for ppc32 on 10.6 PPC too. The issue just concerns Rosetta **and, by the way, ppc64 on Leopard**. The latter still broken completely: https://github.com/iains/darwin-toolchains-start-here/discussions/29

> This has all worked for years

As a matter of fact, Macports for ppc64 will not work without manual trickery. While I am sure you can fix it locally for yourself and likely have done that earlier, it is broken in the master. For years :)

Reply:
when last I built cctools and ld64 universal on 10.5 ppc/ppc64, I recall no troubles. It’s been a while, though, since I did that… not sure what troubles you’re having with ppc64 on 10.5 now. Other than the recent icu won’t build as ppc64 with gcc7, of course, haven’t seen or heard any issues or seen any tickets about issues.

As you’ve been around only a few months, must be hard to know what has been broken For Years :).

Reply:
> when last I built cctools and ld64 universal on 10.5 ppc/ppc64, I recall no troubles. It’s been a while, though, since I did that… not sure what troubles you’re having with ppc64 on 10.5 now. Other than the recent icu won’t build as ppc64 with gcc7, of course, haven’t seen or heard any issues or seen any tickets about issues.

I cannot test `llvm-3.4` for `ppc64` right now (still away from PPC hardware), but last time(s) I tried, it failed – which is why I opened that discussion thread.
None of Pythons will build as ppc+ppc64 from the master. (I have fixed them for myself eventually, but you saw how it went with even simple fixes for Pythons for PPC – with PR being just closed, so I left it at that.)
It is overall impossible to have Macports on Leopard as ppc64-only, as it is.

> As you’ve been around only a few months, must be hard to know what has been broken For Years :).

Unless someone has broken archaic ports recently, I assume either no one ever tested them or otherwise did not push fixes into the master.

Reply:
nobody has ever tested building anything on macports with Rosetta on 10.6, other than one-off jokey things. Nobody would have ever pushed anything into master to support this. I still am 100% against pushing such oddball specific fixes, but you seem to have induced Christopher into doing it for you.

cctools and ld64 build as universal ppc/ppc64 on Leopard PPC, have for years. That is a (more or less) supported MacPorts configuration.

I picture the time you move on from this passion and are no longer interested in building things as rosetta on 10.6, and we start stripping all these hacks out.

Reply:
As before, however -- **proper fixes**, that do the right thing for all builds and in the process also happen to fix 10.6-on-PPC (and Rosetta on 10.6) are completely perfectly fine, and much appreciated.

having llvm-3.4 build with CPU_SUBTYPE_ALL for PPC build may be an OK fix.... Jeremy did the ppc7400 thing for a reason, not sure exactly what it was. Maybe just to build faster code, not sure.

Reply:
@kencu You might have built llvm34 for ppc64 for yourself, what I am saying is that it will not work from portfiles from the master. It gonna fail on llvm34.

Rosetta has a value for testing ppc for those who do not have hardware. Can’t see a reason why not to fix something that can be fixed and doesn’t affect other systems.

Reply:
> Nobody has ever tested building anything on macports with Rosetta on 10.6, other than one-off jokey things. Nobody would have ever pushed anything into master to support this. I still am 100% against pushing such oddball specific fixes, but you seem to have induced Christopher into doing it for you.

Ken, I always welcome your input. And whenever you see something that isn't kosher, please speak up!

You've also provided plenty of positive praise for a number of Sergey's PRs. So honestly, this is a surprising - if not shocking - comment.

Perhaps if you can elaborate a bit, relative to what you consider acceptable, vs. unacceptable?

Reply:
> I picture the time you move on from this passion and are no longer interested in building things as Rosetta on 10.6, and we start stripping all these hacks out.

I'm absolutely fine with everything that has been done thus far, for Sergey's past PRs. And much of the time, the additions are a few simple lines of logic, which are benign for non-PPC/Rosetta builds.

Even in the cases where additional patches are necessary, those are cleanly segregated from the portfile itself. And Not a Big Deal.

So the long and short of it is this: I certainly won't be letting anyone rip these changes out of anything I maintain! And folks can also expect strong pushback, if they try to rip out these changes for other ports... whether they maintain them or not.

Reply:
Unmaintainable spaghetti hacks to support an oddball build system that absolutely nobody except for 1 person ever uses are not something I can possibly support.

I hear the exact same thing coming from others.

But -- have at it if you like it ;> What happens in MacPorts is certainly not up to me.

Reply:
> Rosetta has a value for testing ppc for those who do not have hardware. Can’t see a reason why not to fix something that can be fixed and doesn’t affect other systems.

You keep saying this, but a 10.5 Leopard VM runs just fine on Parallels and is a much better testbed for PowerPC code.

Reply:
> A 10.5 Leopard VM runs just fine on Parallels

Have you achieved installed Parallels tools?

Reply:
Sure, didn't seem to have any troubles -- this was all done about 5 years ago, now, and updated when Parallels updates...

![Screen Sharing Picture September 3, 2022 at 8 06 21 PM PDT](https://user-images.githubusercontent.com/21211439/188295359-2771ee27-4332-40c2-aa4f-1096d55bf692.png)

Reply:
Ken, all my attempts to install Patallels Tools are failed🥺

Reply:
@catap On 10.5.8, not 10.6.8, right?

Reply:
@barracuda156 yes, on 10.5.8; I also have it as Parallels VM. It is officially supported and I've opened an issue inside Parallels to fix tools with provided crashes, gdb dumps and so on :)

Reply:
Heisenbergs I guess. 

I have not yet installed the very latest Parallels update -- I generally hang back a little with things like that. The VM is ripping fast with 8 cores and 16GB of RAM.

Sometimes after I have used it for a few days the screen display will start tearing during refreshes and similar glitches; I log into it using Screen Sharing and it's fine that way, until I reboot it.

I can't get Tiger running on Parallels, but VirtualBox runs Tiger reasonably well.

Reply:
@kencu well, I've successfully boot Tiger in Parallels, but it had a random crashes :)

Reply:
> Ken, all my attempts to install Patallels Tools are failed🥺

Oh, one thing I can tell you here -- sometimes you must manually uninstall Parallels Tools and then reboot, and then install parallels tools fresh. Updating Parallels Tools can fail on older system VMs if it is already installed.

Reply:
> @kencu well, I've successfully boot Tiger in Parallels, but it had a random crashes :)

very interesting. I must have tried a dozen times, and it would never accept the Tiger boot.

Reply:
> very interesting. I must have tried a dozen times, and it would never accept the Tiger boot.

It was a kind of magic. Installer can boot, or can crash. Usually crash but sometimes it may boot. If boot, it may try to install. With a lot of luck it can do it.

Tiger inside parallels seems like a walk in minefield. It randomly crashes. And it is unuseful at all.

Reply:
I left a link to the virtualbox/tiger instructions in your parallels post, but here it is for you again https://github.com/ranma42/TigerOnVBox

Honestly, though, it's a timehole. Tiger is missing some really basic supports like rpath, and that is getting to be a big PITA. I had supported Tiger for some years on MacPorts because I was using an old Tiger PPC box to send out faxes using hylafax for about 15 years and didn't want to go through the project of updating it.

But I moved the hylafax server to a current Ubuntu system with brand new faxmodem PCI cards, and mothballed the Tiger PPC system, converting it to debian11 instead. Now that box runs rust and llvm-14 :>

Reply:
> I had supported Tiger for some years on MacPorts because I was using an old Tiger PPC box to send out faxes using hylafax for about 15 years and didn't want to go through the project of updating it.

sounds like you live in Germany too :) here fax is quite useful in 2022.